### PR TITLE
ENCD-4353-Refactor-out-matrix-and-result

### PR DESCRIPTION
**Pull request by zoldello**
Only functional code changes.

### CODE REVIEW

File: src/encoded/viewconfigs/matrix.py [hunk 1]
```diff
@@ -202,50 +196,60 @@ def preprocess_view(self):
...
+            'filters': [],
+            '@id': matrix_route + self._search_base,
+            '@type': ['Matrix'],
+            'notification': '',
+            'title': self._set_result_title(type_info),
+            'matrix': matrix,
+            'view': self._construct_result_views(type_info),
+        }
+        search_query, used_filters = self._construct_query(search_result, matrix)
+        es_results = self._elastic_search.search(
```
Comment (zoldello) [inline src/encoded/viewconfigs/matrix.py]: You mentioned this in the ticket:

Second Note: Move see how much of the results creation could be moved below the es query search line. https://github.com/ENCODE-DCC/encoded/blob/dev/src/encoded/viewconfigs/matrix.py#L216

There is a Texas standoff situation where es_results needs search_result and search_result needs es_results; breaking this standoff could be tricky and not necessarily pretty. I recommend we leave it.

Comment (caseylitton) [reply src/encoded/viewconfigs/matrix.py]: Because of 'construct_query'?  I suggested looking into to why construct_query needs search_results and how that dependency was impacted by ENCD-4345/SNO-69.  I believe we've already fixed this.

But good job.  This is exactly what I wanted you to find per the Second Note.  A important step in cleaning up the back end matrix code is identifying unnecessary dependencies so we can fix them.  I think this is the line of code that lead us to correcting the set_filters via ENCD4345/SNO-69.

Comment (zoldello) [reply src/encoded/viewconfigs/matrix.py]: I might can move the es_results above search_result. However, I think it would be better if I do that in a different ticket. The changes are tangential to ENCD-4353 (this ticket). 

How important is that to you? Two different tickets perform refactoring on matrix.py and adding a third ticket to the mix is doable if it is important.

Comment (caseylitton) [reply src/encoded/viewconfigs/matrix.py]: The only reason we're handing in 'search_result' to '_construct_query' is because 'set_filters' may add a 'filters' key with a list of terms dicts.  Let's make this dependency more clear by not passing in the final search_results dict.  Just pass in a dict with a 'filters' key.  

This will allow us to move the construction of the final search results down lower in the function.

File: src/encoded/viewconfigs/matrix.py [hunk 2]
```diff
@@ -202,50 +196,60 @@ def preprocess_view(self):
...
+            index=self._es_index
+        )
         aggregations = es_results['aggregations']
         total = aggregations['matrix']['doc_count']
-        self._result['matrix']['doc_count'] = total
-        self._result['matrix']['max_cell_doc_count'] = 0
-        self._result['facets'] = self._format_facets(
+        search_result['matrix']['doc_count'] = total
+        search_result['matrix']['max_cell_doc_count'] = 0
+        search_result['facets'] = self._format_facets(
```
Comment (caseylitton) [inline src/encoded/viewconfigs/matrix.py]: Don't add the facets to the search_results since it won't exist at this point due to other change.

File: src/encoded/viewconfigs/matrix.py [hunk 3]
```diff
@@ -202,50 +196,60 @@ def preprocess_view(self):
...
-            y_groupings + [x_grouping]
+            y_groupings + [x_grouping],
+            matrix
         )
-        self._result['matrix']['y'][y_groupings[0]] = aggregations['matrix'][y_groupings[0]]
-        self._result['matrix']['x'].update(aggregations['matrix']['x'])
-        self._result.update(
+        search_result['matrix']['y'][y_groupings[0]] = aggregations['matrix'][y_groupings[0]]
+        search_result['matrix']['x'].update(aggregations['matrix']['x'])
+        search_result.update(
```
Comment (caseylitton) [inline src/encoded/viewconfigs/matrix.py]: Also don't update search result since it doesn't exist yet.   Just store the actions returned from 'search_result_actions' in a var.

File: src/encoded/viewconfigs/matrix.py [hunk 4]
```diff
@@ -202,50 +196,60 @@ def preprocess_view(self):
...
             search_result_actions(
                 self._request,
                 self._doc_types,
                 es_results
             )
         )
-        self._result['total'] = es_results['hits']['total']
-        if self._result['total']:
-            self._result['notification'] = 'Success'
+        search_result['total'] = es_results['hits']['total']
```
Comment (caseylitton) [inline src/encoded/viewconfigs/matrix.py]: Now that we've collected everything we need to make search_results, build it here with remaining keys.

File: src/encoded/viewconfigs/matrix.py [hunk 5]
```diff
@@ -128,14 +125,14 @@ def _validate_items(self, type_info):
         if msg:
             raise HTTPBadRequest(explanation=msg)
 
-    def _set_query_aggs(self, query):
+    def _set_query_aggs(self, query, search_result, matrix):
```
Comment (caseylitton) [inline src/encoded/viewconfigs/matrix.py]: Let try removing the matrix dependency here too.  Put this in a separate commit than the search_results change in the other comments.  

_set_query_aggs only needs a single list for 
```
field in matrix['x']['facets'] or
                field in matrix['y']['facets']
```
and these two things 
```
        x_grouping = matrix['x']['group_by']
        y_groupings = matrix['y']['group_by']
```
for _construct_xygroupings

File: src/encoded/viewconfigs/matrix.py (hunk 2 above)
Comment (zoldello) [reply src/encoded/viewconfigs/matrix.py]: Let me know if this issue still exist. I refactored but I could not see a reason to move this.

File: src/encoded/viewconfigs/matrix.py (hunk 3 above)
Comment (zoldello) [reply src/encoded/viewconfigs/matrix.py]: Let me know if the principle behind this still exist. I could not see a reason to move it after the refactor

File: src/encoded/viewconfigs/matrix.py (hunk 5 above)
Comment (zoldello) [reply src/encoded/viewconfigs/matrix.py]: I forgot move this to a separate commit. Is ther separate commit important to you?

File: src/encoded/viewconfigs/matrix.py [hunk 6]
```diff
@@ -202,50 +197,68 @@ def preprocess_view(self):
...
-        es_results = self._elastic_search.search(body=query, index=self._es_index)
+        matrix['search_base'] = search_route + self._search_base
+        matrix['clear_matrix'] = matrix_route + '?type=' + self._doc_types[0]
+        matrix_x_y = {
+            'x': matrix['x'],
+            'y': matrix['y'],
+        }
+        search_query, used_filters = self._construct_query(
+            result_filters,
+            matrix_x_y
```
[the remark this answers is not in the document]
Comment (zoldello) [reply src/encoded/viewconfigs/matrix.py]: Yes, you are correct and I normally try to keep name consistent. I used the name "matrix_x_y" in process_view to avoid a name collision in this method with "matrix". However, for methods called by process_view directly or indirectly, like "construct_query", they properly should use "matrix". This also applies to other use of "matrix" in other classes.Do you prefer I used "matrix_x_y" anyway?

[the remark this answers is not in the document]
Comment (caseylitton) [reply src/encoded/viewconfigs/matrix.py]: Which other classes use these MatrixView functions?

File: src/encoded/viewconfigs/matrix.py [hunk 7]
```diff
@@ -76,10 +73,11 @@ def _construct_result_views(self, type_info):
             views.append(self._view_item.summary_report)
         return views
 
-    def _construct_xygroupings(self, query, filters, negative_filters):
+    # pylint: disable=no-self-use
```
Comment (zoldello) [inline src/encoded/viewconfigs/matrix.py]: @caseylitton This is not part of the requirements. However, I added it to silence Codacy (this method out to be a static function as "self" is never used). It is something that needs attention in the future.

Comment (caseylitton) [reply src/encoded/viewconfigs/matrix.py]: Let remove this, the codacy issue it not a problem.  If we do want to fix it the remove self from the arguments and make add the staticmethod decorator.

Comment (zoldello) [reply src/encoded/viewconfigs/matrix.py]: I removed the comment to disable pylint warning. Going forward, I will not rebase while a PR is undergoing review.
